libs/x-content/src/main/java/org/elasticsearch/xcontent/XContentString.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.xcontent;
+
+import java.nio.ByteBuffer;
+
+public interface XContentString {
+    record EncodedBytes(byte[] bytes, int offset, int length) implements Comparable<EncodedBytes> {
+        public EncodedBytes(byte[] bytes) {
+            this(bytes, 0, bytes.length);
+        }
+
+        @Override
+        public int compareTo(EncodedBytes o) {
+            return ByteBuffer.wrap(bytes, offset, length).compareTo(ByteBuffer.wrap(o.bytes, o.offset, o.length));
```

Comment:
Random thought: Is there any value in peeling off a fast-path case for when all the fields are identical by `==`? Not sure whether this ever actually happens, but it would avoid two object allocations.

Reply:
Yeah, essentially you're making sure you're no slower than the build in `Record.equals()` for the case that it returns true. Could be a little slower for the `false` case.